### PR TITLE
Preview

### PR DESCRIPTION
**Pull request by edwinmp**

### CODE REVIEW

File: Gruntfile.js [hunk 1]
```diff
@@ -1,12 +1,18 @@
...
     devtool: false,
-    plugins: plugins
+    plugins: widgetPlugins
+});
+const previewPlugins = webpackConfig[1].plugins.slice();
+previewPlugins.push(new webpack.optimize.UglifyJsPlugin());
+Object.assign(webpackConfigRelease[1], webpackConfig[1], {
+    devtool: false,
+    plugins: previewPlugins
 });
```
Comment (acellam) [inline Gruntfile.js]: -Code above is not DRY. this looks better

`
"use strict";
const webpack = require("webpack");
const webpackConfig = require("./webpack.config");
const webpackConfigRelease = [];

webpackConfig.forEach(function (currentWebpackConfig) {
    const webpackLoaderOptionsPlugin = currentWebpackConfig.plugins.slice(0);
    const configRelease = {};

    webpackLoaderOptionsPlugin.push(new webpack.optimize.UglifyJsPlugin());
    Object.assign(configRelease, currentWebpackConfig, {
        devtool: false,
        plugins: webpackLoaderOptionsPlugin
    });
    webpackConfigRelease.push(configRelease)
});
`

File: src/Carousel.webmodeler.ts [hunk 2]
```diff
@@ -0,0 +1,37 @@
+import { Component, createElement } from "react";
+
+import { Carousel, Image } from "./components/Carousel";
+import CarouselContainer, { CarouselContainerProps } from "./components/CarouselContainer";
+import { Alert } from "./components/Alert";
+
+// tslint:disable-next-line
+const image = require("base64-image-loader!./img/Preview.jpg");
```
Comment (acellam) [inline src/Carousel.webmodeler.ts:8]: i think u dont need this loader because webpack require can already do that

File: Gruntfile.js [hunk 3]
```diff
@@ -71,6 +77,15 @@ module.exports = function (grunt) {
             }
         },
 
+        file_append: {
+            addSourceURL: {
+                files: [ {
+                    append: "\n\n//# sourceURL=Carousel.webmodeler.js\n",
+                    input: "dist/tmp/src/Carousel.webmodeler.js"
```
Comment (Andries-Smit) [inline Gruntfile.js]: Please use the variabele from package.json
`" + pkg.widgetName + "`

File: src/Carousel.xml [hunk 4]
```diff
@@ -1,86 +1,88 @@
...
-            <category>Behavior</category>
-            <description>The page to open when an image is clicked</description>
-        </property>
-    </properties>
-</widget>
+<?xml version="1.0" encoding="utf-8"?>
+<widget id="com.mendix.widget.custom.carousel.Carousel" needsEntityContext="true" offlineCapable="true" xmlns="http://www.mendix.com/widget/1.0/" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.mendix.com/widget/1.0/ ../xsd/widget.xsd">
+    <name>Carousel</name>
+    <description>Carousel displaying images</description>
+    <icon>
```
Comment (Andries-Smit) [inline src/Carousel.xml:5]: icon could be a single line `<icon>long bla bla bas64</icon>`

File: src/ui/Carousel.scss [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+    }
+
+    @mixin on-hover-focus {
+        color: #fff;
+        text-decoration: none;
+        filter: alpha(opacity=90);
+        outline: none;
+        opacity: .9;
+    }
+    &:focus {
```
Comment (Andries-Smit) [inline src/ui/Carousel.scss]: missing newline

File: typings/jasmine.d.ts [hunk 6]
```diff
@@ -1,8 +1,8 @@
+// tslint:disable no-namespace
 declare namespace jasmine {
     interface Any {
         asymmetricMatch(other: any): boolean;
     }
 
-    interface MatchersUtil {}
-    export var matchersUtil: MatchersUtil;
+    export const matchersUtil: MatchersUtil;
```
Comment (Andries-Smit) [inline typings/jasmine.d.ts:7]: 👍

File: webpack.config.js [hunk 7]
```diff
@@ -19,9 +19,9 @@ module.exports = {
     module: {
         rules: [
             { test: /\.ts$/, use: "ts-loader" },
-            { test: /\.css$/, loader: ExtractTextPlugin.extract({
+            { test: /\.scss$/, loader: ExtractTextPlugin.extract({
                 fallback: "style-loader",
-                use: "css-loader"
+                use: "css-loader!sass-loader"
```
Comment (Andries-Smit) [inline webpack.config.js:27]: use array.
```
        use: [ {
                loader: "style-loader" // creates style nodes from JS strings
            }, {
                loader: "css-loader" // translates CSS into CommonJS
            }, {
                loader: "sass-loader" // compiles Sass to CSS
            } ]
```

File: webpack.config.js [hunk 8]
```diff
@@ -41,3 +41,31 @@ module.exports = {
         })
     ]
 };
+
+const previewConfig = {
+    entry: "./src/Carousel.webmodeler.ts",
+    output: {
+        path: path.resolve(__dirname, "dist/tmp"),
+        filename: "src/Carousel.webmodeler.js",
```
Comment (Andries-Smit) [inline webpack.config.js]: can we extract all names out of this file fully?
Just as we did on the grunt.js
```
const package = require("./package");
const widgetName = package.widgetName;
const name = package.widgetName.toLowerCase();


filename: `src/${widgetName}.webmodeler.js`,
```
etc

File: Gruntfile.js (hunk 1 above)
Comment (Andries-Smit) [reply Gruntfile.js]: There are some better options even, but still testing

File: src/Carousel.webmodeler.ts (hunk 2 above)
Comment (edwinmp) [reply src/Carousel.webmodeler.ts:8]: Thought it did, but it doesn't... need the loader.

File: Gruntfile.js (hunk 3 above)
Comment (edwinmp) [reply Gruntfile.js]: Done!

File: src/ui/Carousel.scss (hunk 5 above)
Comment (edwinmp) [reply src/ui/Carousel.scss]: Done.

File: webpack.config.js (hunk 7 above)
Comment (edwinmp) [reply webpack.config.js:27]: Couldn't get that to work.

File: Gruntfile.js (hunk 1 above)
Comment (edwinmp) [reply Gruntfile.js]:
```
"use strict";
const webpack = require("webpack");
const webpackConfig = require("./webpack.config");
const webpackConfigRelease = [ {}, {} ];
webpackConfig.forEach(function (config, index) {
    config.plugins.push(new webpack.optimize.UglifyJsPlugin());
    Object.assign(webpackConfigRelease[index], config, {
        devtool: false,
        plugins: config.plugins
    });
});
```

File: webpack.config.js (hunk 8 above)
Comment (edwinmp) [reply webpack.config.js]: Done!

**Pull request merged**